### PR TITLE
Logwitherr

### PR DESCRIPTION
**Pull request by charliesignalfx**

### CODE REVIEW

File: log/log.go [hunk 1]
```diff
@@ -141,3 +141,28 @@ func IfErr(l Logger, err error) {
...
+func IfErrAndReturn(l Logger, err error) error {
+	if err != nil {
+		l.Log(Err, err)
+	}
+	return err
+}
+
+// IfErrWithKeys logs an error with the supplied logger if it is not nil.
+// The error will be appended to the end of the supplied keys/messages
+func IfErrWithKeys(l Logger, err error, intf ...interface{}) {
```
Comment (jgheewala) [inline log/log.go]: If we use this a lot in the code then won't it have performance impact with append. Can we do a benchmark on this just to get an idea?

File: log/log.go [hunk 2]
```diff
@@ -141,3 +141,28 @@ func IfErr(l Logger, err error) {
...
+	if err != nil && l != nil {
+		l.Log(append(intf, err)...)
+	}
+}
+
+// IfErrWithKeysAndReturn logs an error with the supplied logger if it is not nil and then returns the original error.
+// The error will be appended to the end of the supplied keys/messages
+func IfErrWithKeysAndReturn(l Logger, err error, intf ...interface{}) error {
+	if err != nil && l != nil {
+		l.Log(append(intf, err)...)
```
Comment (jgheewala) [inline log/log.go]: same as above

File: log/log.go (hunk 1 above)
Comment (charliesignalfx) [reply log/log.go]: What would you like for me to compare it to?  IfErr?

**Comment by charliesignalfx** [comment]
It looks like the new log functions that make use of append() have almost a 35% performance hit when compared with similar log functions that do not use the append.

```
BenchmarkIfErr-12                         	 2000000	       748 ns/op
BenchmarkIfErrAndReturn-12                	 2000000	       730 ns/op
BenchmarkIfErrWithKeys-12                 	 1000000	      1131 ns/op
BenchmarkIfErrWithKeysAndReturn-12        	 1000000	      1134 ns/op
```

**Pull request merged**